dtls/server-dtls13-event.c

```diff
@@ -0,0 +1,511 @@
+
+
+    exitVal = 0;
+cleanup:
+    free_resources();
+    wolfSSL_Cleanup();
+
+    return exitVal;
+}
+
+static int newFD(void)
+{
+    int fd;
+    int on = 1;
+    struct sockaddr_in servAddr;        /* our server's address */
+
+    /* Create a UDP/IP socket */
+    if ((fd = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) {
+        perror("socket()");
+        return INVALID_SOCKET;
+    }
+    memset((char *)&servAddr, 0, sizeof(servAddr));
+    /* host-to-network-long conversion (htonl) */
+    /* host-to-network-short conversion (htons) */
+    servAddr.sin_family      = AF_INET;
+    servAddr.sin_addr.s_addr = htonl(INADDR_ANY);
+    servAddr.sin_port        = htons(SERV_PORT);
+
+    if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (char*)&on, sizeof(on)) != 0) {
+        perror("setsockopt() with SO_REUSEADDR");
+        goto cleanup;
+    }
+#ifdef SO_REUSEPORT
+    if (setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, (char*)&on, sizeof(on)) != 0) {
+        perror("setsockopt() with SO_REUSEPORT");
+        goto cleanup;
+    }
+#endif
+    if (fcntl(fd, F_SETFL, O_NONBLOCK) < 0) {
+        perror("fcntl");
+        goto cleanup;
+    }
+
+    /* Bind Socket */
+    if (bind(fd, (struct sockaddr*)&servAddr, sizeof(servAddr)) < 0) {
+        perror("bind()");
+        goto cleanup;
+    }
+    return fd;
+cleanup:
+    if (fd != INVALID_SOCKET) {
+        close(fd);
+        fd = INVALID_SOCKET;
+    }
+    return INVALID_SOCKET;
+}
+
+static int newPendingSSL(void)
+{
+    /* Create the pending WOLFSSL Object */
+    if ((pendingSSL = wolfSSL_new(ctx)) == NULL) {
+        fprintf(stderr, "wolfSSL_new error.\n");
+        return 0;
+    }
+
+    wolfSSL_dtls_set_using_nonblock(pendingSSL, 1);
+
+    if (wolfSSL_SetChGoodCb(pendingSSL, chGoodCb, NULL) != WOLFSSL_SUCCESS ) {
+        fprintf(stderr, "wolfSSL_SetChGoodCb error.\n");
+        return 0;
+    }
+
+    if (wolfSSL_set_fd(pendingSSL, listenfd) != WOLFSSL_SUCCESS) {
+        fprintf(stderr, "wolfSSL_set_fd error.\n");
+        return 0;
+    }
+
+    return 1;
+}
+
+static void newConn(evutil_socket_t fd, short events, void* arg)
+{
+    struct sockaddr_in cliaddr;         /* the client's address */
+    socklen_t          cliLen;
+    char               b;
+    int                ret;
+    int                err;
+    int                drop = 1;
+    /* Store pointer because pendingSSL can be modified in chGoodCb */
+    WOLFSSL*           ssl = pendingSSL;
+
+    (void)events;
+    (void)arg;
+
+    /* Get the incoming address */
+    cliLen = sizeof(cliaddr);
+    ret = (int)recvfrom(listenfd, &b, sizeof(b), MSG_PEEK,
+            (struct sockaddr*)&cliaddr, &cliLen);
+    if (ret <= 0) {
+        perror("recvfrom()");
+        goto error;
+    }
+
+    if (wolfSSL_dtls_set_peer(ssl, &cliaddr, cliLen) != WOLFSSL_SUCCESS) {
+        fprintf(stderr, "wolfSSL_dtls_set_peer error.\n");
+        goto error;
+    }
+
+    drop = 0;
+    ret = wolfSSL_accept(ssl);
```

Comment:
So do you think that returning a `WANT_READ` after processing an incorrect msg is the correct path (if using non-blocking)? This way we give the user a chance to change the peer address before the next `recvfrom` is called.

Reply:
Setting it in the IO would be the best solution but I wanted to write the examples to use the default IO. Also resetting the peer after sending the HRR looks like it will be confusing for users. I think by definition this feature is not supported for blocking sockets using the default IO callbacks (unless you have an idea how this could work?).

Reply:
If the `listenMode` is on we can set it also in the default IO if we document that I think. 

I think this should be supported for blocking socket as well (after all, one can use event-driven paradigm with blocking socket). I feel that setting the peer on a SSL that is supposed to reply to every packet is wrong. 

Setting the peer (because we need a destination to send HRR to) and resetting it right after it may be a solution, not sure if the best one though. 

I don't feel that returning WANT_READ and moving the burden of this logic to the user is a good choice.

Just to list another alternative, you can also think to hide this logic by a "wolfSSL_dtls_listen" kind of new API. It may require some more effort.

Reply:
What do you think of getting the peer from `recvfrom` in `wolfio.c` when we don't have a confirmed peer and the user hasn't set the peer? This way the peer selection logic is in the network IO layer and we cut down on the number of system calls (no peek necessary).

Reply:
This is what I was thinking of but I would rather make it explicit (getting the peer when wolfSSL_dtls_listen_mode() was invoked or something with a better name) than relying on the fact that the peer isn't set yet.
The peer selection logic can't be in the network IO layer only because it needs to be un-setted after sending the HRR, but I think this isn't a problem.

Reply:
After thinking about this again, how would you write an event driven server with blocking sockets? You would need to have a thread for each connection you make because otherwise the blocked socket would stall other connections. So for me, I don't see how it is possible to write a server for multiple simultaneous connections in one thread.

Reply:
yes, you are right, we can't write an event-driven with blocking socket with wolfssl

Reply:
The peer is now set in `EmbedReceiveFrom`. This simplifies the server to the point where `newConn` only calls `wolfSSL_accept` and does error handling.